Quantized sigmoid function

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32335 Benchmarks for quantized LSTM (static)
* #32074 Static quantization for LSTM (front)
* #31934 Static quantization for LSTM (back)
* **#31851 Quantized sigmoid function**

Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Quantized sigmoid function

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 809130cf:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=809130cfb20fa59ed7699bbbe61bb94b35d6eda7).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4127685) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80130304) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80130304))
<details>
<summary>
<code>Jan 04 01:33:01 /pytorch/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp:64:38: error: ‘mobile_pthreadpool’ is not a member of ‘caffe2’</code>
</summary>

```
Jan 04 01:33:01                  from /pytorch/aten/src/ATen/cpu/vec256/vec256.h:5, 
Jan 04 01:33:01                  from /pytorch/aten/src/ATen/native/cpu/Loops.h:34, 
Jan 04 01:33:01                  from /pytorch/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp:5: 
Jan 04 01:33:01 /pytorch/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp:16:17: error: aggregate ‘at::native::qsigmoid_stub at::native::qsigmoid_stub’ has incomplete type and cannot be defined 
Jan 04 01:33:01  DEFINE_DISPATCH(qsigmoid_stub); 
Jan 04 01:33:01                  ^ 
Jan 04 01:33:01 /pytorch/aten/src/ATen/native/DispatchStub.h:153:43: note: in definition of macro ‘DEFINE_DISPATCH’ 
Jan 04 01:33:01  #define DEFINE_DISPATCH(name) struct name name 
Jan 04 01:33:01                                            ^~~~ 
Jan 04 01:33:01 /pytorch/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp: In function ‘at::Tensor at::native::qnnpack_sigmoid(at::Tensor)’: 
Jan 04 01:33:01 /pytorch/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp:64:38: error: ‘mobile_pthreadpool’ is not a member of ‘caffe2’ 
Jan 04 01:33:01    pthreadpool_t threadpool = caffe2::mobile_pthreadpool(); 
Jan 04 01:33:01                                       ^~~~~~~~~~~~~~~~~~ 
Jan 04 01:33:01 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mkldnn/BinaryOps.cpp.o 
Jan 04 01:33:03 [ 56%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/mkldnn/Conv.cpp.o 
Jan 04 01:33:04 gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/quantized/cpu/qsigmoid.cpp.o] Error 1 
Jan 04 01:33:04 gmake[2]: *** Waiting for unfinished jobs.... 
Jan 04 01:33:13 gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Jan 04 01:33:13 gmake: *** [all] Error 2 
Jan 04 01:33:13 Traceback (most recent call last): 
Jan 04 01:33:13   File "setup.py", line 755, in <module> 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,63 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+        return at::quantize_val<scalar_t>(output_scale, output_zero_point,
+                                          value_dy);
+      },
+      [&](Vec value_qx) -> Vec {
```
The code in 231-236 and 238-244 are doing the same operations, should one of them be deleted?

jamesr66a replied:
The first is the scalar body, the second is the vector body. They're not the same

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,63 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+  auto zero_point_vec = Vec256<float>((float)zero_point);
+  auto scale_neg_zp_premul_vec = scale_vec * zero_point_vec.neg();
+
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "qsigmoid", [&]() {
```
Can you also add benchmark results showing time spent on different operations within an LSTM cell to understand what fraction is spent in sigmoid?

z-a-f replied:
That will be a separate set of PRs during the optimization

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

supriyar commented on aten/src/ATen/native/quantized/cpu/qsigmoid.cpp:
```diff
@@ -0,0 +1,89 @@
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_sigmoid_nc_q8(
+    sigmoid_op,
+    input_contig.size(0) /* batch size */,
+    (uint8_t*)input_contig.data_ptr<c10::quint8>() /* input data */,
```
@raghuramank100 qnnpack expects uint8 input and produces uint8 result. So no dequantize should be necessary here.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1865,6 +1894,30 @@ def test_qnnpack_tanh(self, X):
+    @given(X=hu.tensor(shapes=hu.array_shapes(1, 5, 1, 5),
+                       qparams=hu.qparams(dtypes=torch.quint8)))
+    def test_qnnpack_sigmoid(self, X):
+        # Note: In QNNPACK the output scale and zero_point can only be
```
Suggest adding a test that sweeps the input over the range [-4,4] with a scale of 8/256 and a zero-point of 128 to ensure that we check that the LUT implementation in qnnpack and the FBGEMM implementation are identical in numerics. Alternately,  we could also implement FBGEMM qnnpack with a LUT based approach.

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

jamesr66a commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -193,6 +193,63 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+        auto value_dx = value_qx.dequantize(scale_vec, zero_point_vec,
+                                            scale_neg_zp_premul_vec);
+        for (int idx = 0; idx < value_dx.size(); ++idx) {
+          value_dx[idx] = Vec256<float>(0.0f) - value_dx[idx];
```
Can't you use `.neg()` here? negation should be an xor with sign bit, which is 1 cycle latency, v.s. 3-4 cycle latency for sub

jamesr66a reviewed (commented):
Looks pretty good to me, just one small comment

supriyar commented on test/test_quantized.py:
```diff
@@ -82,6 +82,7 @@ def pool_output_shape(input_size, kernel_size, padding, stride,
         output_size += 1
     return output_size
 
+
```
nit: Extra line

z-a-f replied:
Oh, I think this line has to be here -- isn't it python style to have two blank lines before class defs?

supriyar reviewed (approved):
LGTM!

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)

Commit:
Update on "Quantized sigmoid function"


Differential Revision: [D19280716](https://our.internmc.facebook.com/intern/diff/D19280716)
